### PR TITLE
QVAC-24114 fix: fail the load when the model cannot decode

### PR DESCRIPTION
**Pull request by simon-iribarren**
## 🎯 What problem does this PR solve?

`loadModel` returning a model id does not mean the model can run. Measured on a 24 GiB M4 Pro with Gemma 4 31B Q4_K_M (18.25 GiB):

- `ctx_size 1024, gpu_layers 99`: loads in 25 s (Metal over-commits past `recommendedMaxWorkingSetSize`), then the **first** completion fails with `[TextLlm] failed to decode next token`, and every later request fails with `evalMessageWithTools: failed to decode input tokens`.
- `ctx_size 1024, gpu_layers 48`: same — loads, cannot decode.
- `device 'cpu'`: decodes (0.1 tok/s), so the GGUF is fine; this is a memory-placement failure surfacing at the worst possible time, on the first user request, with no way for the app to tell a usable load from a dead one.

Two existing layers fail to catch it: fabric's `common_init_from_params` warmup **discards the warmup decode status**, and `LoadFitNormalization` disables warmup outright (`params.warmup = false`).

This is also the necessary complement to projection-based fitting (QVAC-24112 / tetherto/qvac-fabric-llm.cpp#214): macOS memory pressure is bistable, so a configuration can pass every static budget and still die at decode. Only an actual decode proves usability.

## 📝 How does it solve it?

After context creation in `LlamaModel`, run one strict BOS/EOS probe decode (mirroring fabric's warmup shape, encoder models included) and **check its status**:

- failure → clear the probe's KV trace, throw a structured `UnableToLoadModel` StatusError; no model id is handed out.
- success → the probe doubles as a warmup: weights faulted, Metal pipelines compiled, so first-token latency improves as a side effect.
- skipped for finetuning, which never serves inference from this context.

No API change; the SDK surfaces the load error through the existing path.

## 🧪 How was it tested?

- Compiles clean via `bare-make` on macOS arm64.
- `run-cpp-addon-tests` and `run-desktop-addon-tests` are applied, so the addon suites run on the CI runners.
- Local runtime verification blocked by a pre-existing local-toolchain issue (see note at top): a **clean, unmodified** build of current `main` fails identically before `initFromConfig`, so the published-prebuild control confirms the harness and isolates the local build as the variable.
- Regression fixture for a runner with ≥ Gemma-class models: load Gemma 4 31B Q4_K_M at `ctx_size 1024, gpu_layers 99` on a 24 GiB Apple silicon machine → `load()` must reject; Qwen3.5 0.8B/9B and gpt-oss-20B fixtures must keep loading (plus one warmup decode).

### CODE REVIEW

File: packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp [hunk 1]
```diff
@@ -79,6 +79,72 @@ LlamaModel::LlamaModel(
...
+      decoderStart = bos;
+    }
+    tokens.assign(1, decoderStart);
+  }
+  if (status == 0 && llama_model_has_decoder(model)) {
+    status = llama_decode(
+        lctx,
+        llama_batch_get_one(
+            tokens.data(), static_cast<int32_t>(tokens.size())));
+  }
```
Comment (iancris) [inline packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp:135]: *The probe is a near-verbatim copy of fabric's warmup block with one clamp removed, and the assertion it thereby trips is `abort()`, not a catchable error.*

Upstream's warmup decodes a **clamped** batch (`common/common.cpp:1447`):

```cpp
llama_decode(lctx, llama_batch_get_one(tmp.data(), std::min(tmp.size(), (size_t) params.n_batch)));
```

Here `std::min(..., params.n_batch)` is gone. The engine's batch-size check is an assertion, not an error return — `GGML_ASSERT(n_tokens_all <= cparams.n_batch)` — and `GGML_ASSERT` expands to `GGML_ABORT` (`ggml/include/ggml.h:288`). So this does not throw `UnableToLoadModel`; it terminates the process, with nothing for the SDK to catch.

`batch_size` is public and documented (`docs/continuous-batching.md:62` constrains only `batch_size >= parallel`), nothing floors `params.n_batch`, and `cparams.n_batch = min(n_ctx, params.n_batch)`. So `batch_size: 1` with `parallel: 1` builds a 2-token BOS/EOS probe, compares `2 <= 1`, and aborts the host process at load. `kMinNCtx = 8` bounds `n_ctx`, not `n_batch`. The encode branch has the same exposure against `GGML_ASSERT(cparams.n_ubatch >= n_tokens)`, so `ubatch_size: 1` aborts on encoder models.

**Suggested fix:** restore upstream's clamp on both calls, so the probe can never exceed the configured batch geometry. A one-token probe is equally diagnostic, so nothing is lost.

```cpp
status = llama_decode(
    lctx,
    llama_batch_get_one(
        tokens.data(),
        static_cast<int32_t>(std::min<size_t>(tokens.size(), params.n_batch))));
```

This needs `params` — or `llama_n_batch(lctx)` / `llama_n_ubatch(lctx)` — passed into the helper.

File: packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp [hunk 2]
```diff
@@ -230,6 +296,15 @@ void LlamaModel::init(bool acquireLock) {
...
       std::move(llamaInit));
 
+  // Finetuning drives its own graph and never serves inference from this
+  // context, so the inference probe would only add noise there.
+  if (snap->llmContext_ && !params.training) {
+    probeDecodeOrThrow(
+        snap->llmContext_->getCtx(),
+        snap->llmContext_->getModel(),
+        errorWhenFailed);
+  }
```
Comment (iancris) [inline packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp:338]: *A probe failure leaves `batchScheduler_` null while `parallel >= 2` keeps admitting concurrent jobs, so they decode on one `llama_context` unserialised.*

The probe throws after `snap->llmContext_ = createContext(...)` publishes the context (`:294`) but before `cacheManager_.emplace()` (`:308`) and `initBatchScheduler()` (`:314`). Nothing resets `llmContext_`, so the state stays installed with `batchScheduler_ == nullptr`.

That null is the sentinel the request path reads as "single worker, no locking needed": the guard against concurrent jobs racing the shared context is itself gated on the scheduler existing (`:703`, *"without a scheduler the single path is the only worker and stays safe"*), and `:716` routes everything to `processSinglePath()`, which takes no exclusive lock. But `maxConcurrency` is parsed once from `config["parallel"]` at `createInstance` (`AddonJs.hpp:386-390`) and never re-derived, so the job layer keeps admitting N jobs — N threads decoding on one `llama_context`, racing its KV cache and sampler. `supportsBatching()` (`:1188-1191`) also ignores `batchScheduler_`, so it reports `true` and the pre-admission gate at `AddonJs.hpp:449-455` passes.

The reachable path is the post-finetune reload: `LlamaFinetuner.cpp:441-445` swallows the reload exception and rethrows the original finetuning error, so the JS layer sees a failed finetune rather than a failed load, and never unloads. `setInitLoader` has already replaced `state_` (`:190-191`), so a model that was serving correctly ends up here.

**Suggested fix:** make a failed probe leave no usable state, so the existing "`llmContext_` null ⇒ not loaded" invariant still holds.

```cpp
if (snap->llmContext_ && !params.training) {
  try {
    probeDecodeOrThrow(
        snap->llmContext_->getCtx(),
        snap->llmContext_->getModel(),
        errorWhenFailed);
  } catch (...) {
    snap->llmContext_.reset();
    throw;
  }
}
```

File: packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp [hunk 3]
```diff
@@ -79,6 +79,72 @@ LlamaModel::LlamaModel(
...
+    status = llama_decode(
+        lctx,
+        llama_batch_get_one(
+            tokens.data(), static_cast<int32_t>(tokens.size())));
+  }
+
+  // Leave no trace of the probe regardless of outcome.
+  llama_memory_clear(llama_get_memory(lctx), true);
+  llama_synchronize(lctx);
+  llama_perf_context_reset(lctx);
```
Comment (iancris) [inline packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp]: *The cleanup wipes the KV buffers before waiting for the decode that is still writing to them, which is an unsynchronized concurrent write on every load.*

`llama_decode` does not synchronize — the graph is submitted with `ggml_backend_sched_graph_compute_async` and the readback uses `ggml_backend_tensor_get_async`, with the trailing `synchronize()` commented out. `llama_memory_clear` does not either: `llama_kv_cache::clear(data)` calls `ggml_backend_buffer_clear`, which waits only for itself — on host-visible Vulkan (Mali/Android) a CPU `memset` over mapped memory with no GPU sync, and on CUDA a `cudaMemsetAsync` on `cudaStreamPerThread`, unordered against the backend's non-blocking streams. Only `llama_synchronize` → `ggml_backend_sched_synchronize` waits on all backends.

So the clear races the decode it is cleaning up after. The ordering is copied verbatim from upstream's warmup (`common.cpp:1449-1450`), but that block is dead code here — `LoadFitNormalization.cpp:843` sets `params.warmup = false` unconditionally — so this PR is what makes the hazard execute, on every load.

The blast radius is limited, since the cells are being reset anyway and stale bytes get overwritten. Worth fixing because it is a genuine data race on GPU-visible memory on every load, and this codebase already knows the rule: `LlamaModel.cpp:1115-1117` and `TextLlmContext.cpp:508-510` both force `llama_synchronize` for exactly this reason.

**Suggested fix:** swap the first two calls so the wait happens before the wipe. `llama_perf_context_reset` must stay last, because `llama_context::synchronize` is what commits `t_eval_us`/`n_queued_tokens`.

```cpp
llama_synchronize(lctx);
llama_memory_clear(llama_get_memory(lctx), true);
llama_perf_context_reset(lctx);
```

File: packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp [hunk 4]
```diff
@@ -79,6 +79,72 @@ LlamaModel::LlamaModel(
...
+    throw qvac_errors::StatusError(
+        ADDON_ID,
+        error,
+        string_format(
+            "%s: model loaded but cannot decode (probe decode failed with "
+            "status %d); the resolved configuration does not fit this "
+            "device's usable memory\n",
+            __func__,
+            status));
+  }
```
Comment (iancris) [inline packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp:161]: *Every non-zero decode status is reported as a memory-fit failure, but `llama_decode` documents four distinct failure classes and only some concern memory.*

`llama.h` documents the return contract:

```
//    1 - could not find a KV slot for the batch (try reducing the size of the batch or increase the context)
//    2 - aborted     (processed ubatches will remain in the context's memory)
//   -1 - invalid input batch
// < -1 - fatal error (processed ubatches will remain in the context's memory)
```

Only `< -1` is the backend-compute failure this PR targets. `-1` is returned for `n_tokens == 0`, a failed `balloc->init`, a backend-sampling violation and a pooled-embedding mismatch — none of them memory conditions — and the probe can produce `-1` itself, via the `bos` fallback in the encoder branch when a model declares neither a decoder-start token nor a BOS.

The misdiagnosis propagates: `benchmarks/performance/llm-parameter-sweep.js:285` and `:489` classify any error whose text contains `UnableToLoadModel` as a VRAM error and skip the case. A probe failing with `-1` would be recorded as a memory-capacity result for a model whose problem was never memory.

The status is already carried in the message, so the fix is just to stop asserting a cause the code has not established.

**Suggested fix:** describe the status rather than naming one cause, so a caller can tell a capacity problem from a malformed batch.

```cpp
string_format(
    "%s: model loaded but cannot decode (probe decode failed with status %d; "
    "1 = no KV slot, -1 = invalid batch, <= -2 = allocation or compute failure)\n",
    __func__,
    status));
```

File: packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp (hunk 2 above)
Comment (iancris) [inline packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp:338]: *The finetune success path calls `reload()` unguarded, so a probe failure converts a finished, already-saved finetune into a thrown `UnableToLoadModel`.*

`LlamaFinetuner.cpp:424` calls `model_.reload(FinetuneConfigOverrides{})` on the success and pause path. `active` is false there, so `params.training` is false (`LoadFitNormalization.cpp:844`) and the probe runs. The ordering is what makes this a correctness problem rather than a latency one:

- `:415` — `saveLoraAdapter(...)`, so the adapter is already on disk.
- `:423` — `status` is already `"COMPLETED"` or `"PAUSED"`.
- `:424` — `model_.reload(...)`, not wrapped, throws.
- `:425` — `return status` never runs.

The exception propagates out of `finetune()` into `LlamaModel.cpp:588`, where the return value becomes the job's terminal result — so the caller is told their training run failed while the adapter it produced sits on disk.

The error path immediately below wraps the identical call defensively, precisely so a failing reload cannot mask the original error:

```cpp
try {
  model_.reload(FinetuneConfigOverrides{});
} catch (...) {
  QLOG_IF(Priority::ERROR, "Failed to reload model after finetuning error");
}
throw;
```

The success path needs the same treatment for the same reason. It also cascades: the throw from `:424` is caught at `:426`, which reaches `:441` and reloads *again* — a second full weights load and probe — before rethrowing. `reload()`'s doc comment documents no failure semantics at all.

**Suggested fix:** wrap the `:424` call as the error path already does, or return the status with a flag noting the context could not be returned to inference mode — so a completed finetune is never reported as a load failure.

**Review by gianni-cor — CHANGES_REQUESTED** [review]
if the model can't decode better to check the cause then this change

File: packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp (hunk 1 above)
Comment (simon-iribarren) [reply packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp:135]: Fixed in 850ddf941 exactly as suggested: both calls now clamp to the context's configured geometry — `llama_n_batch(lctx)` on decode, `llama_n_ubatch(lctx)` on encode — with a floor of 1, so the probe can never exceed what the engine asserts on. Thanks for catching that the assertion is an abort, not a return; that would have been a crash on a perfectly legal `batch_size: 1` config.

File: packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp (hunk 2 above)
Comment (simon-iribarren) [reply packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp:338]: Fixed in 850ddf941: the probe call is wrapped, and on failure the published context, cache manager and batch scheduler are all reset before rethrowing — restoring the "llmContext_ null ⇒ not loaded" invariant so nothing downstream can read a half-initialised state as "single worker, no locking". Went one step past the suggestion by also resetting cacheManager_/batchScheduler_, since neither may outlive the context they point at.

File: packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp (hunk 3 above)
Comment (simon-iribarren) [reply packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp]: Fixed in 850ddf941: `llama_synchronize` now runs first, then `llama_memory_clear`, with `llama_perf_context_reset` kept last since synchronize is what commits the counters it clears. You are right that this PR is what made the upstream ordering execute at all — noted in the code comment, along with the two in-repo precedents you cited.

File: packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp (hunk 4 above)
Comment (simon-iribarren) [reply packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp:161]: Fixed in 850ddf941 with your exact message shape: the error now describes the status classes (1 = no KV slot, -1 = invalid batch, <= -2 = allocation/compute) instead of asserting a memory cause. Good catch on the benchmark sweep classifying on message text — that was the concrete harm.

File: packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp (hunk 2 above)
Comment (simon-iribarren) [reply packages/llm-llamacpp/addon/src/model-interface/LlamaModel.cpp:338]: Fixed in 850ddf941: the success-path reload is now guarded like the error path — a completed/paused finetune returns its status and logs that the model could not be returned to inference mode, instead of throwing a load error over a run whose adapter is already on disk (and no longer cascades into the error path's second reload).

**Comment by simon-iribarren** [comment]
@gianni-cor fair point — short version of why we did both:

**We did find the cause, and it's fixed where it belongs.** The pre-load fit check was deciding with bad inputs: macOS reports GPU memory *per-process* (so our check, which runs in a helper process, always saw an empty machine no matter what was loaded), and the CPU-side gauge literally reported all memory as free. [qvac-fabric-llm.cpp#214](https://github.com/tetherto/qvac-fabric-llm.cpp/pull/214) fixes those inputs, so the "will it fit" verdicts now track reality much more closely. That's the cause-side fix.

**But we also measured that prediction alone can never be 100% right here.** Near the memory ceiling, macOS sometimes compresses/swaps its way out and the load works fine — and sometimes it doesn't. We have a real config (Gemma 31B, 48/61 layers on GPU) that passes even the *corrected* budget and still dies at decode. No formula separates those two cases; only actually decoding a token does.

So the division of labour is:

- fabric #214 + the advisory fit (#4010): predict before loading, with honest numbers → catches most bad loads cheaply.
- this PR: the safety net for what prediction can't call → generate one throwaway token at load time, and if that fails, fail the load right there with a clear error.

Without this, the failure surfaces as `[TextLlm] failed to decode next token` on the user's **first message** — after the app was told the load succeeded. Today nothing decodes between load and that first message (fabric's warmup discards its own decode status, and we disable warmup anyway).

All of @iancris's findings are addressed in 850ddf941. Happy to gate the probe behind a load option if you'd prefer it opt-in.
